### PR TITLE
Ngcc es5 resolving

### PR DESCRIPTION
**Pull request by petebacondarwin**
This will replace #25048 when #24897 has landed in master.
Ignoring the fixup commits at the end: the penultimate commit is from #25048 - and the final commit is the meat of this PR.

### CODE REVIEW

File: packages/compiler-cli/src/ngcc/src/host/esm5_host.ts [hunk 1]
```diff
@@ -54,6 +55,29 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
     }
   }
 
+  /**
+   * Parse a function declaration and to find the relevant metadata about it.
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngcc/src/host/esm5_host.ts]: _and to_ --> _and_

File: packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts [hunk 2]
```diff
@@ -406,6 +406,23 @@ const EXPORTS_FILES = [
...
+    }
+    function bar(x) {
+      if (x === void 0) { x = 42; }
+      return x;
+    }
+    function complex() {
+      var x = 42;
+      return 42;
+    }
+  `
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts]: I think it is also worth having tests like these (to test some aspects of looking for param initializers):

```js
function baz() {
  var x;
  if (x === void 0) { x = 42; }
  return x;
}

var x;
function qux() {
  if (x === void 0) { x = 42; }
  return x;
}
```

**Review by gkalpak — APPROVED** [review]

File: packages/compiler-cli/src/ngcc/src/host/esm5_host.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/src/host/esm5_host.ts]: or *and to* --> *to*

File: packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts (hunk 2 above)
Comment (gkalpak) [reply packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts]: Great tests you added there 👍 
I still think the tests in my previous comment are worth adding too :grin:

Comment (gkalpak) [reply packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts]: The sentiment is great. It's just that your tests and my tests test different code paths.
(We need both.)

Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts]: So many code paths!

Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/test/host/esm5_host_spec.ts]: Done

**Comment by gkalpak** [comment]
Rebased on master (now that #24897 has been merged) and force-pushed.

**Comment by trotyl** [comment]
Sorry for an off-topic question, I wonder if current `ngcc` architecture will break use case of https://github.com/angular/angular/issues/24580 which doesn't use `esm` at all?

Previously `compiler-cli` needs only `.metadata.json` and `.d.ts`, rather than `.js`, making arbitrary JavaScript processing possible.

**Comment by petebacondarwin** [comment]
@trotyl thanks for linking this. I am pretty sure that Ngcc will not be able to consume the modified UMD file which is all that appears to be left after "securing" a project.

That being said, I believe that going forward, one would be able to republish the original secured files after they have been compiled with ngtsc

**Review by alxhub — APPROVED** [review]

**Comment by alxhub** [comment]
@trotyl good point, and the answer as @petebacondarwin says is indeed no, it won't work with such packages. `ngcc` only operates on Angular Package Format packages.

File: packages/compiler-cli/src/ngcc/src/host/esm5_host.ts [hunk 3]
```diff
@@ -54,6 +55,29 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
...
+   * @param node the function declaration to parse.
+   */
+  getDefinitionOfFunction<T extends ts.FunctionDeclaration|ts.MethodDeclaration|
+                          ts.FunctionExpression>(node: T): FunctionDefinition<T> {
+    const parameters =
+        node.parameters.map(p => ({name: getNameText(p.name), node: p, initializer: null}));
+    let lookingForParamInitializers = true;
+
+    const statements = node.body && node.body.statements.filter(s => {
+      lookingForParamInitializers =
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngcc/src/host/esm5_host.ts]: OK

**Comment by trotyl** [comment]
@alxhub @petebacondarwin Thanks for reply, I'm fine with that, if finally everything will be on `ngtsc`, it's reasonable for not wasting effort.

**Comment by petebacondarwin** [comment]
Superceded by #25090

**Pull request closed without merging**